Add "Collapse/Expand all" actions to instruments panel

MaxiSchwindler opened the pull request:
Resolves: #19562 <!-- Replace `NNNNN` with a GitHub issue number, or a direct link if the issue is not on GitHub -->

<!-- Add a short description of and motivation for the changes here -->

<!-- Replace `[ ]` with `[x]` to fill the checkboxes below -->

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

RomanPudashkin commented:
@MaxiSchwindler please fix codestyle. Thanks!

MaxiSchwindler commented:
> @MaxiSchwindler please fix codestyle. Thanks!

Hm, the thing that makes the codestyle job fail is the line
```
action.title = expand ? TranslatableString("instruments", "Expand all instruments")
                      : TranslatableString("instruments", "Collapse all instruments");
```
which it wants to replace with
```
action.title = expand ? TranslatableString("instruments", "Expand all instruments")
               : TranslatableString("instruments", "Collapse all instruments");
```
I would argue that the first option is nicer, and I think I've seen it in other places in MuseScore as well... but I can change it of course

Jojo-Schmitz commented:
Don't argue with `uncrustify` ;-)

avvvvve commented:
Hi @MaxiSchwindler, nice work here! I have a couple requests that will make this airtight: 

1. If you click "Collapse all instruments" when scrolled to the bottom of a long list of instruments, the panel may appear empty until you scroll again inside the instruments panel (see video). Instead, when collapsing all, if the list is longer than the panel height, the bottom instrument should appear at the bottom of the window. Perhaps this requires recalculating the height of the list before refreshing the view?

https://github.com/musescore/MuseScore/assets/20806406/914a0d5c-6860-4aed-8709-1feb1e554882

2. The Collapse/Expand actions are only available in the main score. They should also be available when viewing parts.
Note that instruments that have never been unhidden in parts have a special state where in addition to being hidden, they do not show a dropdown arrow or gear icon. These instruments should not be expanded when using the "Expand all instruments" action.

![image](https://github.com/musescore/MuseScore/assets/20806406/acfe1bf8-2903-4a10-888d-e1461ee11e10)

cbjeukendrup commented:
About the code style error: in situations like that one, I tend to choose the following compromise, which `uncrustify` does like:
```c++
    action.title = expand 
                   ? TranslatableString("instruments", "Expand all instruments")
                   : TranslatableString("instruments", "Collapse all instruments");
```

MaxiSchwindler commented:
@avvvvve 
1) Does this always happen for you? I was only able to reproduce that once, every other time it worked as expected... I'm on Linux, if that makes a difference.
2) The part about the action not affecting the instruments without the arrow is done in the latest commit (11bffe0). However, I'm not sure how to correctly go about showing the context menu for parts as well:

The reason why these two actions are not in the context menu for parts are these lines in InstrumentsPanelContextMenuModel::load():

```cpp
if (!m_masterNotation || !notation || m_masterNotation->notation() != notation) {
    clear();
} else {
    loadItems();
}
```
The menu containing the "Instrument ordering" item and the expand/collapse items is only built in buildMenu(), which is called in loadItems(). However, when the current notation is not the master notation (i.e. we are looking at a part?), the `m_masterNotation->notation() != notation` condition is true, the menu is cleared, and loadItems is never called. I'm not quite sure what the reason is; I guess the "Instrument ordering" item should only appear in the main score?
Should this now be more of a three state thing? I.e 
```cpp
if (!m_masterNotation || !notation ) {
    clear();
} else if (m_masterNotation->notation() != notation){
   // build menu without "instrument ordering" but with "expand/collapse"
}else {
    loadItems(); 
}
```
? Maybe @RomanPudashkin can help? :)

cbjeukendrup commented:
@MaxiSchwindler Regarding point 2, yes, that seems a good analysis and a good solution!

About point 1, it might be a macOS thing, since I think I could reproduce it as well. I was thinking, since TreeView inherits ScrollView, you could use ScrollView's [`flickableItem` property](https://doc.qt.io/qt-5/qml-qtquick-controls-scrollview.html#flickableItem-prop) to obtain the underlying Flickable, and then call that Flickable's [`returnToBounds()` method](https://doc.qt.io/qt-5/qml-qtquick-flickable.html#returnToBounds-method).
So, in `expandCollapseAll`, you'd add `instrumentsTreeView.flickableItem.returnToBounds()`, or if that doesn't work, `Qt.callLater(instrumentsTreeView.flickableItem.returnToBounds)`. I haven't tested that yet, though.

MaxiSchwindler commented:
So, I added the three-way-if, now the expand/collapse is in both main score and part score menus, and the instrument ordering only in the main score. I also added the `instrumentsTreeView.flickableItem.returnToBounds();` call in the qml file; I think it _should_ fix the problem, but as I can't reproduce it, I can't really say for sure... Maybe someone with MacOS could quickly test it?

cbjeukendrup commented on src/instrumentsscene/qml/MuseScore/InstrumentsScene/InstrumentsPanel.qml:
```diff
@@ -166,6 +170,21 @@ Item {
+                            }
+                        }
+                    }
+                    instrumentsTreeView.flickableItem.returnToBounds();
```
@MaxiSchwindler After some experimenting, I found that this seems to work: 
```suggestion
                    flickable.returnToBounds();
```
Why it works: as I just remembered, we've disabled the default scroll functionality from TreeView and put it inside a Flickable ourselves. So we need to return _that_ Flickable to its bounds!

It seems that the behaviour will still not be perfect when you scroll manually after using "collapse all", but that seems really a Qt bug, so that's probably the best we can get.

MaxiSchwindler commented:
@avvvvve 

Issue 1 should hopefully be fixed now thanks to @cbjeukendrup's suggestion!

For issue 3, The problem was that the menu was only updated on a notation change, but not in load() itself; I changed that in e783e98.

cbjeukendrup commented on src/instrumentsscene/view/instrumentspanelcontextmenumodel.cpp:
```diff
@@ -83,11 +94,57 @@ void InstrumentsPanelContextMenuModel::loadItems()
         currentOrder.customized = false;
         m_orders.push_back(currentOrder);
     }
+}
```
Now that the functionality of `loadItems()` has changed so much, I think it would be nice to rename it, for example to `loadInstrumentOrders()`.

MaxiSchwindler replied:
done :)

cbjeukendrup commented:
I think that would make sense. Otherwise the only way to get it fixed is if someone with macOS finds a fix and posts it here. So we can as well merge the PR and have that someone fix it themselves.

cbjeukendrup commented:
In that case I'll merge it right now! I don't think it needs to be tested by the QA engineers right now, because @avvvvve has tested it enough and there is going to be a follow-up anyway.

cbjeukendrup commented:
Thanks @MaxiSchwindler!

Eism commented:
@MaxiSchwindler, if you have free time, please take a look at https://github.com/musescore/MuseScore/issues/20411

MaxiSchwindler commented:
> @avvvvve Sadly I still cannot reproduce this behaviour on Linux... therefore I don't think I will be able to fix that. Would it be possible to merge this pull request and create a follow-up issue for your issue, such that someone with MacOS can work on it?

@Eism As said above, I sadly can't reproduce it on Linux... that's why we decided to merge this as-is and directly open #20411
